Add Embedders\Word2Vec, Graph\Trees\Heap, and Word2Vec Unit Test files

RichDavis1 opened the pull request:
This is the initial push for the Word2Vec embedding class as well as additional dependencies.

Commit:
Add Embedders\Word2Vec, Graph\Trees\Heap, and Word2Vec Unit Test files

This is the initial push for the Word2Vec embedding class as well as additional dependencies.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,983 @@
+     *
+     * @var int   
+     */		
+	protected int $rand_multiplier = 4294967296;		
```
Until we end support for PHP 7.2 and 7.3, we do not use 7.4 typed properties

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,983 @@
+     *
+     * @var int[]     
+     */		
+	protected $raw_vocab = array();
```
Please use [] instead of array() and camelCase variable names for consistency

```php
protected $rawVocab = [];
```

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,983 @@
+     *
+     * @return bool
+     */
+    public function trained() : bool
```
This can be simplified to

```php
return !empty($this->vectors);
```

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,983 @@
+     */		
+	public function train(Dataset $dataset) : void	
+	{
+		DatasetIsNotEmpty::check($dataset);
```
Is this a tab? Did you run ```$ composer fix``` and ```$ composer check```?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,983 @@
+		$min_alpha   = .0001;
+		$start_alpha = $this->alpha;
+
+		foreach(range(0, ( $this->epochs - 1)) as $i){	
```
```foreach (range(...) as $i) {``` works differently in Python than it does in PHP ... This line will actually create and allocate memory for the array produced by range which makes it slower and less efficient than ```for ($i = 0; $i < $this->epochs; ++$i) {```.

Refer this comment https://www.php.net/manual/en/function.range.php#85331

Commit:
Edit Word2Vec Class

Removed typed properties from word2vec class.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,955 @@
+		int $epochs       = 10,
+		int $minCount     = 2
+	) {	
+
```
Extra whitespace ... I believe ```composer fix``` will automatically fix this

Commit:
Edit word2vec class

Implementing suggested edits.

Commit:
Merge remote-tracking branch 'upstream/master'

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+ */
+class Word2Vec implements Embedder
+{
+    use LoggerAware;
```
Is this a Verbose (implementing the Verbose interface) Embedder?

If so, it should implement the Verbose interface, if not this trait is unneeded.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+        if ($this->logger) {
+            $this->logger->info('Embedder init ' . Params::stringify($this->params()));
+
+            $this->logger->info('Computing high-dimensional affinities');
```
This looks like it's left over from t-SNE

Do the concept of high-dimensional affinities exist in Word2Vec?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+     *
+     * @var int[]
+     */
+    protected $raw_vocab = [];
```
camelCase for variable names

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+     *
+     * @var \Tensor\Vector
+     */
+    protected $neu1e;
```
What is the rationale behind this variable name?

RichDavis1 replied:
That's what was used as the error variable in the w2v.py file. I used most of the original variable names from that file to make it easy for me to follow along.

andrewdalpino replied:
If you were to name this variable, what would you name it?

RichDavis1 replied:
Most likely just $error.

RichDavis1 replied:
My comment seems to have disappeared, I would rename it $error.

andrewdalpino commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+class Heap
+{
+    /**
+     * An array of sanitized and exploded sentences.
```
Is this the correct description for this variable?

RichDavis1 replied:
No it is not, good eye.

andrewdalpino commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+
+        $max_range = (($n / 2) - 1);
+
+        foreach (array_reverse(range(0, $max_range)) as $i) {
```
This can be more efficient (see previous comment)

```php
for ($i = $maxRange; $i > 0; --$i) {
```

RichDavis1 replied:
It would be but hopping back to your previous statements regarding making this efficient after launching it, since this is a one-time step that's not involved in the training loop, the incremental time isn't too big of a deal.

I can update however if you'd like.

RichDavis1 replied:
Does the '--$i' reverse the $i variable in the iteration sequence?

andrewdalpino replied:
`--$i` pre decrements the $i variable ... in the case of that for loop $i is counting backward by 1 from $maxRange

You could even do ..

```php
foreach (range($max_range, 0) as $i) {
```

and it would avoid the array_reverse (an unnecessary O(n) operation), however the for loop is optimal in memory as well as time so let's use that unless you can think of a good reason not to.

andrewdalpino commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+     */
+    public function heappop() : ?array
+    {
+        if (!is_array($this->heap) || empty($this->heap)) {
```
For consistency can you use 'or' instead of '||' and 'and' instead of '&&' unless we need to explicitly denote precedence

https://www.geeksforgeeks.org/and-vs-as-operator-in-php/

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+use Rubix\ML\Other\Helpers\Params;
+use Rubix\ML\Other\Traits\LoggerAware;
+
+use Tensor\Matrix;
```
No spaces between class imports, however, do add spaces between a blocks of class imports, function imports, and constant imports.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+     * @var int
+     */
+    protected $rand_multiplier = 4294967296;
+    /**
```
Single blank line between class properties

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+        $this->preprocess($sentences);
+        $this->prepareWeights();
+
+        $min_alpha = .0001;
```
Should this be a class constant? Also, for consistency, make sure to add a leading zero

RichDavis1 replied:
I believe this and randMultiplier could work as a class constant for now.

andrewdalpino replied:
Let's try and abstract constants as much as possible

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+        $start_alpha = $this->alpha;
+
+        for ($i = 0; $i < $this->epochs; ++$i) {
+            $cur_epoch = $i;
```
Is this variable assignment necessary?

RichDavis1 replied:
Not any longer with the new for loop.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+    }
+
+    /**
+     * Scanning vocab, preparing vocab, sorting vocab, setting sampling methods
```
Single blank line between class method description and its annotations

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+            $prepped_sentence = $this->prepSentence($sentence);
+
+            $this->corpus[] = $prepped_sentence;
+            $words = array_merge($words, $prepped_sentence);
```
This merge operation may be suboptimal ... we repay the cost to reindex the $words array each call

RichDavis1 replied:
What would you recommend?

Keep in mind, this is only pre-processing, and only occurs once. So the gains from optimization would be limited compared to a process involved in epoch iteration.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+     */
+    private function prepSentence(string $sentence) : array
+    {
+        $sentence = (string) preg_replace('#[[:punct:]]#', '', strtolower($sentence));
```
Just a suggestion, but it may be helpful to define these regex patterns as class constants

RichDavis1 replied:
I'm kind of opposed to defining these two patterns as constants. There are two of them and I don't think someone would see the constant definition and immediately identify their use case. They are also shorter definitions.

I'd be happy to make them constants if you'd like but I feel would take attention away from the definitions that matter.

andrewdalpino replied:
What name would you give such a constant?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+     */
+    private function scanVocab(array $words) : void
+    {
+        foreach ($words as $i => $word) {
```
Unused counter $i

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+     */
+    private function createCumTable() : void
+    {
+        $ns_exponent = 0.75;
```
Should this be a class constant?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+            $train_words_pow += pow($this->vocab[$this->index2word[$word_index]]['count'], $ns_exponent);
+        }
+        
+        foreach (range(0, ($this->vocab_count - 1)) as $word_index) {
```
Marking this as it's suboptimal vs a for loop (see previous comments)

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+        }
+        
+        foreach (range(0, ($this->vocab_count - 1)) as $word_index) {
+            $cumulative += pow($this->vocab[$this->index2word[$word_index]]['count'], $ns_exponent);
```
Can we rewrite this using the ** operator instead of the pow() function for consistency and to avoid the function call?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+     */
+    private function prepareWeights() : void
+    {
+        foreach (range(0, ($this->vocab_count - 1)) as $i) {
```
Flagging loop

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+            $this->vectors[$i] = Vector::rand($this->dimensions)->subtractScalar(0.5)->divideScalar($this->dimensions);
+        }
+
+        $this->neu1e = Vector::quick(array_fill(0, $this->dimensions, 0));
```
This can be simplified to ...

```php
$this->neu1e = Vector::zeros(0, $this->dimensions);
```

RichDavis1 replied:
I get this result when I update. Dimensions are set to 100 in my example.

![image](https://user-images.githubusercontent.com/63970045/80890706-ab8df380-8c84-11ea-96f9-546c394b732f.png)

RichDavis1 replied:
Interestingly enough, I don't believe I encountered this issue when updating the other embedWord method to do the same thing.

andrewdalpino replied:
My bad that code was for a matrix ... should be

```php
$this->neu1e = Vector::zeros($this->dimensions);
```

Vector only needs one argument https://github.com/RubixML/Tensor/blob/master/src/Vector.php#L69

RichDavis1 replied:
When I use Vector::zeros($this->dimensions), I get the following fatal error in logs:

```Fatal error: Uncaught InvalidArgumentException: The number of elements must be greater than 0, 0 given. in C:\xampp\htdocs\rubix\RubixML\vendor\rubix\tensor\src\Vector.php:72 Stack trace: #0 C:\xampp\htdocs\rubix\Extras\src\Embedders\Word2Vec.php(605): Tensor\Vector::zeros(0, 100) #1 C:\xampp\htdocs\rubix\Extras\src\Embedders\Word2Vec.php(320): Rubix\ML\Embedders\Word2Vec->prepareWeights() #2 C:\xampp\htdocs\rubix\test.php(359): Rubix\ML\Embedders\Word2Vec->train(Object(Rubix\ML\Datasets\Unlabeled)) #3 {main} thrown in C:\xampp\htdocs\rubix\RubixML\vendor\rubix\tensor\src\Vector.php on line 72```

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+    private function prepareWeights() : void
+    {
+        foreach (range(0, ($this->vocab_count - 1)) as $i) {
+            $this->syn1[] = Vector::quick(array_fill(0, $this->dimensions, 0));
```
This can be simplified to ...

 $this->syn1[] = Vector::zeros(0, $this->dimensions);

RichDavis1 replied:
See other comment.

RichDavis1 replied:
Fatal error: Uncaught InvalidArgumentException: The number of elements must be greater than 0, 0 given. in C:\xampp\htdocs\rubix\RubixML\vendor\rubix\tensor\src\Vector.php:72 Stack trace: #0 C:\xampp\htdocs\rubix\Extras\src\Embedders\Word2Vec.php(605): Tensor\Vector::zeros(0, 100) #1 C:\xampp\htdocs\rubix\Extras\src\Embedders\Word2Vec.php(320): Rubix\ML\Embedders\Word2Vec->prepareWeights() #2 C:\xampp\htdocs\rubix\test.php(359): Rubix\ML\Embedders\Word2Vec->train(Object(Rubix\ML\Datasets\Unlabeled)) #3 {main} thrown in C:\xampp\htdocs\rubix\RubixML\vendor\rubix\tensor\src\Vector.php on line 72

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+    {
+        $word_vocabs = [];
+        $rand = lcg_value();
+        //$rand = (rand() / getrandmax()));
```
Which method was faster? And by how much?

RichDavis1 replied:
Haven't tested yet, let it be in order to push. Is it ok if we test after launching it?

andrewdalpino replied:
Yes, let's use rand() for now since that's what we use everywhere else. I'm interested in knowing which method is faster.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+        //$rand = (rand() / getrandmax()));
+
+        foreach ($sentence as $w) {
+            $vocab_item = $this->vocab[$w] ?? false;
```
Is the $w here meant to denote a word of the sentence?

If so, would $word be a better choice to name the variable?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+
+        foreach ($word_indices as $index) {
+            $this->syn1[$index] = $this->syn1[$index]->addVector($c->rowAsVector($count));
+            $count += 1;
```
We typically use ++$count; when incrementing an integer counter by 1

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+    private function propagateHidden(Matrix $l2, Vector $l1) : Vector
+    {
+        $prod_term = $l1->matmul($l2->transpose());
+        $sigmoid = new Sigmoid();
```
Should the Sigmoid activation function be newed up every time this function is called or should it be a class property?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+        $samples = $dataset->samples();
+        $embeddedDataset = [];
+
+        foreach ($samples as $i => $featureSet) {
```
Unused variable $i

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+        $wordEmbedding = $this->wordVec($word);
+
+        if (!$wordEmbedding) {
+            $wordEmbedding = Vector::quick(array_fill(0, $this->dimensions, 0));
```
Vector::zeros(0, $this->dimensions);

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+        int $epochs = 10,
+        int $minCount = 2
+    ) {
+        if (!$this->acceptedLayer($layer)) {
```
This function can be eliminated and the validation logic can live closer to the constructor

```php
if (!in_array($layer, ['neg', 'hs'])) {
   throw ....
}
```

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(
+        string $layer = 'neg',
```
Is the choice of negative sampling vs hierarchical softmax the most important hyper-parameter or are there other more important hyper-parameters that should occupy this real-estate?

RichDavis1 replied:
It depends on how you look at it but I'd be happy to arrange in the sequence that's partial to the rest of the site: dimensions, alpha, epochs, minCount, window, samplerate if you'd like.

andrewdalpino replied:
How would you rank the hyper-parameters in terms of importance to the user? Are any of the hyper-parameters related and should be grouped?

RichDavis1 replied:
If the focus was importance to user, I'd probably stick with the order w2v has.

My personal order would be window, dimensions, epochs, 'layer', alpha, samplerate, mincount.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+        }
+
+        if ($window > 5) {
+            throw new InvalidArgumentException('Window must be less than 5, $window given.');
```
$window will not be expanded when used within single quotes

https://www.php.net/manual/en/language.types.string.php

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+    {
+        DatasetIsNotEmpty::check($dataset);
+
+        $sentences = $dataset->column(0);
```
What happens if the dataset contains more columns? What if the data in the first column is invalid?

There should be more validation here and if the intent is to only support one vocabulary, it should be known to the user by way of an exception message

RichDavis1 replied:
This kind of deals with what we discussed the other day: a user trying to embed on a multi-feature dataset. 

The only check I can think of is if the contents of the feature are of type string.

andrewdalpino replied:
We're going to need to come up with a way to handle this in order for it to be compatible with the Embedder/Transformer interface

andrewdalpino replied:
So we still need to implement a way to use (or at least handle) *all* columns of the dataset. For example, the user may have multiple text columns.

RichDavis1 replied:
But what if one of the features are not text or they do not want it embedded?

RichDavis1 replied:
There shouldn't be multiple features per sample in a word2vec dataset. This is set up to process one feature of sentences. I can add that in the documentation. But if we want to change it then we need to have a serious conversation. That would require a lot of work and isn't how word2vec is used.

andrewdalpino replied:
Since we decided to only support one text feature, we need to throw an exception if the user tries to fit a dataset with more than 1 column in the samples.

RichDavis1 replied:
Already handled.

andrewdalpino reviewed (changes requested):
Hey @RichDavis1 

We'll start with the comments I left and then drill deeper as we go

Commit:
Edit Word2vec.php

Updated most variables to camel casing, fixed minor invalid spacing, couple of other hot fixes.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,937 @@
+     * @param bool $use_norm
+     * @return \Tensor\Vector $result
+     */
+    public function embedWord(string $word, bool $use_norm = true) : Vector
```
After thinking about it, this method should probably return just a plain PHP array instead of a Vector object. It's leaking an implementation detail

RichDavis1 replied:
This is a situation where I'd have to use range and create an array of zeros, are you ok with this?

andrewdalpino replied:
I'm not sure how that applies here. Can you explain?

RichDavis1 replied:
I can't remember the specific situation regarding this one tbh.

It currently returns a vector of zeros if empty.

This is to work in tandem with returning the word vector if word exists in vocab.

RichDavis1 replied:
I remembered, Vector::zeros($this->dimensions) causes this fatal error. 

Fatal error: Uncaught InvalidArgumentException: The number of elements must be greater than 0, 0 given. in C:\xampp\htdocs\rubix\RubixML\vendor\rubix\tensor\src\Vector.php:72 Stack trace: #0 C:\xampp\htdocs\rubix\Extras\src\Embedders\Word2Vec.php(605): Tensor\Vector::zeros(0, 100) #1 C:\xampp\htdocs\rubix\Extras\src\Embedders\Word2Vec.php(320): Rubix\ML\Embedders\Word2Vec->prepareWeights() #2 C:\xampp\htdocs\rubix\test.php(359): Rubix\ML\Embedders\Word2Vec->train(Object(Rubix\ML\Datasets\Unlabeled)) #3 {main} thrown in C:\xampp\htdocs\rubix\RubixML\vendor\rubix\tensor\src\Vector.php on line 72

Commit:
Edit Word2Vec.php

Made several minor updates per suggestions.

Commit:
Merge remote-tracking branch 'upstream/master'

RichDavis1 committed:
Edit Word2vec, Heap

Switch foreach to for loops, clean naming conventions, additional cleanups per requests.

RichDavis1 committed:
Merge remote-tracking branch 'upstream/master'

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,922 @@
+     * @throws \RuntimeException
+     */
+    public function transform(array &$samples) : void
+    {
```
You are not mutating the `$samples` array passed by reference here, instead you are creating an `$embeddedDataset` array that doesn't get assigned to anything outside the method context.

Each sample in the samples array should have its categorical/string feature columns replaced with its word embedding

https://www.php.net/manual/en/language.references.pass.php

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,920 @@
+                    $embeddings[] = $this->embedWord($word);
+                }
+
+                $sample[$i2] = Matrix::stack($embeddings)->transpose()->mean();
```
This is replacing the feature column at `$i2` (probably should be called `$j` since it's the second inner loop) with a column vector. We do not store vectors (or other arrays for that matter) in the feature columns of a sample matrix. Instead, concatenate the word vectors produced by W2V with the non-categorical features of the sample making sure to unset the old text blob feature column. A sample should contain continuous or categorical scalar features.

Here is a good example

https://github.com/RubixML/RubixML/blob/master/src/Transformers/WordCountVectorizer.php#L193

RichDavis1 committed:
Implemented Embedder & Stateful interfaces in Word2Vec

Implemented Embedder & Stateful interfaces in Word2Vec as well as added/updated necessary contracts.

RichDavis1 committed:
Updated word2vectest

Updated old test methods for new contracts implemented in Word2Vec.

RichDavis1 committed:
Removed transform method unit test.

Removed transform method unit test for the time being until we can discuss how you'd prefer testing be handled.

RichDavis1 committed:
Forgot to run phpcs

fixed styles in line with repo.

RichDavis1 committed:
Edit Word2Vec

Camel cased underscored variables.

RichDavis1 committed:
Edit Word2vec

Removed unnecessary comments.

RichDavis1 committed:
Edit Word2Vec

Updated transform method to update $samples in place.

RichDavis1 committed:
Edit Word2Vec

Ran phpcs

RichDavis1 committed:
Edit Word2Vec

Fixed nested array, and unset $sample indicies once word has been embedded for transform.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+        foreach ($samples as &$sample) {
+            foreach ($sample as $i2 => $sentence) {
+                $preppedSentence = $this->prepSentence($sentence);
+
```
Liking your expressive variable names here

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+            }
+
+            if (isset($embeddings)) {
+                $sample = Matrix::stack($embeddings)->transpose()->mean()->asArray();
```
What happens if the sample contains non text (categorical) feature columns? We should check if the column is categorical (see [DataType](https://github.com/RubixML/RubixML/blob/master/src/DataType.php) value object) and only then perform an embedding on that column. If it is not categorical we should leave it alone.

With this line, we overwrite the entire sample with the embeddings, but what if there were continuous features that didn't get embedded in the original sample?

andrewdalpino replied:
If it's easier we could start off just by handling categorical columns

RichDavis1 replied:
Word2vec is only function for categorical. That's what I've made the checks for. You can verify this in the unit tests.

I will add a check here for post-training usage of the embeddings. That's smart.

RichDavis1 replied:
So the issue here is that as you specified the embedding needs to represent the whole sample. If the user passes a dataset with two features: one categorical and one continuous, and we transform the categorical one, then the output is something like this:
```Array
(
    [0] => Array
        (
            [0] => Array
                (
                    [0] => 0.073936899237247
                    [1] => -0.042769248861282
                    [2] => 0.034288473849103
                    [3] => 0.060570758676996
                    [4] => -0.024365352807662
                    [5] => -0.054537101301627
                    [6] => 0.055912118777915
                    [7] => 0.031113309232784
                    [8] => -0.0096870220834407
                    [9] => -0.06652673585504
                    [10] => -0.052036883331531
                    [11] => -0.093465290156762
                    [12] => 0.030409747492496
                    [13] => -0.0096153259705508
                    [14] => -0.013845659508576
                    [15] => 0.062429033122279
                    [16] => 0.0053496953549063
                    [17] => 0.00032735591163119
                    [18] => 0.050059298965258
                    [19] => -0.065644277009544
                    [20] => 0.0079409481574074
                    [21] => -0.0064926737253694
                    [22] => 0.066788881237452
                    [23] => -0.049248678907282
                    [24] => -0.05025864938639
                    [25] => -0.035897928441286
                    [26] => 0.038975411369952
                    [27] => -0.0022506152860845
                    [28] => 0.042090504347821
                    [29] => -0.12761993703862
                    [30] => -0.0090590375355276
                    [31] => -0.035349723865985
                    [32] => 0.012337233750355
                    [33] => -0.040580268550874
                    [34] => -0.0041515985497862
                    [35] => 0.025362251794008
                    [36] => -0.068469250416346
                    [37] => -0.027960388868277
                    [38] => -0.06985514110827
                    [39] => -0.00076115599053589
                    [40] => 0.013475052392355
                    [41] => -0.012628627348292
                    [42] => -0.13815271233708
                    [43] => -0.025849885597172
                    [44] => 0.040817919010622
                    [45] => -0.02957413229303
                    [46] => -0.051486383479868
                    [47] => 0.0005044815828805
                    [48] => 0.037440586825384
                    [49] => -0.067289972864957
                    [50] => 0.038811874646212
                    [51] => -0.04873065631799
                    [52] => -0.029608522540381
                    [53] => -0.077456147809795
                    [54] => -0.013606250838333
                    [55] => 0.0276403165806
                    [56] => 0.014720418843273
                    [57] => -0.015215311104515
                    [58] => -0.066081697322383
                    [59] => 0.13196366260123
                    [60] => -0.016153006268673
                    [61] => -0.023942195153488
                    [62] => -0.12382280099448
                    [63] => 0.029921574985138
                    [64] => 0.056505968536882
                    [65] => -0.093788057660408
                    [66] => 0.016929875988217
                    [67] => 0.019100403501269
                    [68] => 0.12119411722786
                    [69] => 0.029031812670099
                    [70] => -0.001043484372054
                    [71] => 0.059535551075836
                    [72] => -0.030580417948013
                    [73] => 0.09430545719158
                    [74] => 0.056276469323681
                    [75] => 0.057810823666968
                    [76] => -0.082259321310291
                    [77] => -0.0030038032354214
                    [78] => -0.0091859435424184
                    [79] => -0.0034285290760318
                    [80] => 0.063184130266227
                    [81] => -0.069491852384475
                    [82] => 0.05854247102446
                    [83] => 0.023571855663125
                    [84] => -0.021945659780986
                    [85] => -0.030767950730222
                    [86] => -0.013436536287871
                    [87] => 0.031080757642827
                    [88] => -0.043132133447636
                    [89] => -0.011724875841158
                    [90] => 0.051889978525437
                    [91] => 0.0052859735206233
                    [92] => -0.013598265208169
                    [93] => -0.093086444513104
                    [94] => -0.026665716548919
                    [95] => -0.054662739035899
                    [96] => -0.058788723743905
                    [97] => 0.03912185414767
                    [98] => 0.028593616476334
                    [99] => 0.04251252702093
                )

            [1] => 1
        )
)```

andrewdalpino replied:
Per our convo we'll only support a single categorical feature column at first

RichDavis1 replied:
Per our updated convo, we will support multi-feature categorical columns.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+     *
+     * @var float
+     */
+    protected const MIN_ALPHA = 0.0001;
```
I am just curious, do you know why a lower bound is enforced on the regularization term?

RichDavis1 replied:
Yes, it's more so a target than enforcement. The alpha decay is linear and we use this target as a constant in the calculation for the next epochs alpha.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+        int $epochs = 10,
+        int $minCount = 2
+    ) {
+        if (!in_array($layer, ['neg', 'hs'])) {
```
Did you decide to inject the training method? If so we can add a subfolder to the embedders directory/namespace to contain these objects. We'll need to think of a good name for the interface - any ideas off the top of your head?

What methods would such an interface need to provide Word2Vec?

RichDavis1 replied:
I need to look at it but yes, I think we should abstract the training methods. I will push this with the next release. 

As for the name of the interface, as we discussed, there really isn't one. We will have to go generic. I believe gensim struggled with this as well which is why they simply wanted boolean for which training method.

To put the differences in layman's terms, hierarchical softmax adds info about each word's Huffman encoding, negative sampling builds a cumulative distribution table for drawing random words.  

In my honest opinion, people designate these as _TrainingMethods_ and it would be most intuitive to label the interface as such.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    public function __construct(
+        string $layer = 'neg',
+        int $window = 2,
+        int $dimensions = 5,
```
I think this should be the first argument to the constructor as it depends highly on factors that vary between use-cases such as the size of the user's corpus. It would also be consistent with other Transformers that do embedding/dimensionality reduction.

RichDavis1 replied:
You want the dimensions to be the first hyper-parameter? There are 3 in the snippet, just making sure :)

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+     */
+    public function __construct(
+        string $layer = 'neg',
+        int $window = 2,
```
It's unfortunate but I think we should find another name for this argument if possible. I have been using the window argument canonically to mean 'the number of epochs to consider when determining an early stop of an iterative algorithm."

Would something like `$tokenWindow` or `$wordFrame` work? A better name come to mind?

RichDavis1 replied:
I don't think so. I believe most users are pretty used to 'window' as that's what's used in word2vec. They'd be confused seeing something else. We could do 'windowSize'.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    /**
+     * Training one epoch from the corpus and updating all respective word vectors.
+     */
+    private function trainEpochSg() : void
```
What does the SG stand for here?

RichDavis1 replied:
Skip gram.

andrewdalpino replied:
Since this is the skip gram version of Word2Vec is it necessary to include the abbreviation 'Sg' which can be equivocal for some developer?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+        }
+
+        if ($useNorm) {
+            $result = $this->vectorsNorm[$this->vocab[$word]['index']];
```
Do we need this assignment or can we return right away from within the if statement?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    }
+
+    /**
+     * Preparing and setting an initial raw vocab array for additional preprocessing
```
Not a big deal but we tend to stay away from participles in our comments and just stick to the verb. Example 'Prepare and set an initial raw vocab ...'

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+        }
+
+        $originalUniqueTotal = count($retainWords) + $dropUnique;
+        $retainUniquePct = ((count($retainWords) * 100) / $originalUniqueTotal);
```
Are the outer parenthesis needed or can we reduce some cognitive load?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+        $retainUniquePct = ((count($retainWords) * 100) / $originalUniqueTotal);
+
+        $originalTotal = $retainTotal + $dropTotal;
+        $retainPct = (($retainTotal * 100) / $originalTotal);
```
Are these outer parenthesis needed?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+        } elseif ($this->sampleRate < 1) {
+            $thresholdCount = $this->sampleRate * $retainTotal;
+        } else {
+            $thresholdCount = ($this->sampleRate * (3 + sqrt(5)) / 2);
```
Can `(3 + sqrt(5)) / 2)` be turned into a class constant?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+            $thresholdCount = ($this->sampleRate * (3 + sqrt(5)) / 2);
+        }
+
+        return $thresholdCount;
```
Looks like we can return early from the if statement again ... unless you feel we need the expressiveness of this variable (I do not).

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    /**
+     * An array containing each word in the corpus and it's respective index, count, and multiplier.
+     *
+     * @var mixed[]
```
Does this vocab array contain 3-tuples (index, count, multiplier)? If so does it have a vocabulary for each categorical feature column? Also if so, the type-specification in the docblock should be something more like `@var array[]`.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    /**
+     * An array of each word in the corpus for preprocessing purposes.
+     *
+     * @var int[]
```
An array of each *word* in the corpus yet the type is specified as an array of integers. Is this a mistake?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    protected $vocab = [];
+
+    /**
+     * An array containing each word in the corpus at it's respective index.
```
Why do we store the index here as well as in the `$vocab` array? Do we need both or can we get rid of one to save on memory?

RichDavis1 replied:
The snippet shows the $vocab array so a little confused. Do you mean the index2word? This is because in vocab we store sample_int if subsampling rate is provided, the index for heap if hs is true, count, and the word itself. The vocab is also in a different order than index2word is.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    protected $index2word = [];
+
+    /**
+     * An array containing the hidden layer.
```
This description isn't quite accurate ... it contains a vector of the weights associated with the hidden layer. Also what is the rationale for the property name? Does `syn1` stand for anything? Can we make it more expressive?

RichDavis1 replied:
Syn stands for syntactic and comes from the original Google word2vec.c release. We use this because we perform syntactic/semantic NLP processes with the trained vectors.The names have kind of stuck.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    protected $vectors = [];
+
+    /**
+     * The error vector used during training.
```
Is it needed after training? If not can we unset it *after* training so save memory?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    protected $vectorsNorm = [];
+
+    /**
+     * The cumulative distrubtion table for negative sampling.
```
*distribution*

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    protected $negLabels;
+
+    /**
+     * The training method determined by the layer they determine.
```
Can you clarify this description? I having trouble understanding what you are trying to convey. Who/what is they?

RichDavis1 replied:
This is whether they chose hierarchal softmax or negative sampling.

andrewdalpino replied:
Can you communicate that in this comment for the user?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    protected $window;
+
+    /**
+     * Vector dimension size.
```
Can this be made more informative such as "The dimensionality of each embedded feature column."

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+    protected $sampleRate;
+
+    /**
+     * The amount of L2 regularization applied to the weights of the output layer.
```
Is L2 regularization also applied to the weights of the hidden layer? If so, this  description is not entirely true.

RichDavis1 replied:
It is not.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+     */
+    public function fitted() : bool
+    {
+        return !empty($this->vectors);
```
In the case that a user fits a dataset containing no text features (probably a mistake by the user but we have to still handle the situation) this would not be true. Instead, we could set initial state of the variable to null and check for a non-null value here instead.

RichDavis1 replied:
That's what it was originally set to but you requested that I change it :) 

Just to be clear, this will return false if the model has not been trained. So if they pass empty data, then it will return false.

RichDavis1 replied:
I added a 
`SamplesAreCompatibleWithTransformer::check($dataset, $this);` in fit method which will keep vectors null if they provided non text features.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+        }
+
+        foreach ($samples as &$sample) {
+            foreach ($sample as $i2 => $sentence) {
```
What is the meaning of `$i2` here and is it different from 'the offset of the feature column per sample?'

RichDavis1 replied:
This is just index 2. We have to nest the iteration because of how the samples are constructed.

andrewdalpino replied:
It looks like `$i2` is not used in its context. Can we remove it if so?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,924 @@
+        $l2 = Matrix::stack($this->vectorsNorm);
+        $dists = $mean->transpose()->matmul($l2->transpose())->asArray()[0];
+
+        arsort($dists, 1);
```
Can we use the constant here instead of an integer to be more expressive?

https://www.php.net/manual/en/function.sort.php

andrewdalpino reviewed (changes requested):
Next round, looking better

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Add SoftmaxApproximation interface, HierarchicalSoftmax subclass, NegativeSampling subclass

Abstracted SofmaxtApproximator processes to new SoftmaxApproximation interface, adding compatibility tests, addressed several requested changes.

Commit:
Edit HierarchicalSoftmax.php, edit NegativeSampling.php, edit Word2Vec.php

Handled some last minute clean up, slight optimizations when building the cumulative distribution table.

Commit:
Edit Word2Vec.php

Added in categorical check for transform method.

andrewdalpino commented on src/Embedders/SoftmaxApproximations/SoftmaxApproximation.php:
```diff
@@ -0,0 +1,35 @@
+<?php
```
Generally, interfaces indicate what an object *does* rather than *is* ... for that reason I'd recommend changing the name of this interface to SoftmaxApproximator

I added to the contributing guide here https://github.com/RubixML/RubixML/blob/master/CONTRIBUTING.md#naming

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+        int $epochs = 10,
+        int $minCount = 2
+    ) {
+        if ($window > 5) {
```
Not a big deal, but I generally like to validate constructor arguments in the order that they appear in the constructor. This reduces cognitive load on the developer.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+    public function params() : array
+    {
+        return [
+            'layer' => $this->approximation,
```
Also not a huge deal but since these parameters are used in all sorts of logging and display, it would be nicer for the end-user for these to be in  the same order as they appear in  the constructor.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+     * @param string $sentence
+     * @return string[] $exploded
+     */
+    private function prepSentence(string $sentence) : array
```
What is the reason we cannot use tokenizers to tokenize the words in the sentence?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+                $downsampleTotal += $v;
+            }
+
+            $this->vocab[$w]['sample_int'] = round($wordProbability * (2 ** 32));
```
Is the parenthesis necessary here since exponentiation is performed before multiplication?

PEMDAS

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+        $original = $this->vocab;
+
+        $count = array_column($original, 'count');
+        array_multisort($count, SORT_DESC, $original);
```
Nice usage of multisort

RichDavis1 replied:
Thanks.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+    private function learnHidden(array $wordIndices, Vector $g, Vector $l1) : void
+    {
+        $c = $g->outer($l1);
+        $count = 0;
```
What is the reason for the counter variable? Can we use `$i` as a counter in this case?

```php
foreach ($wordIndices as $i=> $index) {
```

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+     */
+    public function __toString() : string
+    {
+        return 't-SNE {' . Params::stringify($this->params()) . '}';
```
This should be 'Word2Vec' not t-SNE

RichDavis1 replied:
Oops!

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+     * @param array[] $wordVocabs
+     * @return array[]
+     */
+    private function sgSubset(int $pos, array $wordVocabs) : array
```
Do we need the sg prefix here?

RichDavis1 replied:
I believe so yes. If we decide to expand beyond skip-gram then maintaining this would be smart.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+     *
+     * @var int
+     */
+    protected $vocabCount;
```
It looks like this variable isn't set upon instantiation ... should this be `@var int|null`?

RichDavis1 replied:
In the event that it's not set, it should return a 0 in my opinion. This is still int. Adding the nullable also makes things trickier for static analysis. The cleanest way to handle is int with 0 for no vocab.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+     *
+     * @return int[]
+     */
+    public function index2word() : array
```
What value does this public method provide the end user? What use-cases would the user need this?

RichDavis1 replied:
This method is used by the NegativeSampling subclass, line 67.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+     */
+    public function wordVec(string $word, bool $useNorm = true) : ?Vector
+    {
+        if (!array_key_exists($word, $this->vocab)) {
```
Is there a reason `!isset($this->vocab[$word])` cannot be used here? It's a little faster than array_key_exists.

Will an entry in  the vocab ever be null?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+     * @param int $top
+     * @return string[] $result
+     */
+    public function mostSimilar(array $positive, array $negative = [], $top = 20) : array
```
Is there a reason why `$top` is not typehinted in the method signature?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+        $l2 = Matrix::stack($this->vectorsNorm);
+        $dists = $mean->transpose()->matmul($l2->transpose())->asArray()[0];
+
+        arsort($dists, SORT_REGULAR);
```
Is it necessary to use the `SORT_REGULAR` flag here if it's implied by default?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+    {
+        $wordIndices = $this->approximation->wordIndices($predictWord);
+        $l2 = $this->layerMatrix($wordIndices);
+        $fa = $this->propagateHidden($l2, $l1);
```
I've noticed the variables `$l1` and `$l2` used quite often, are these referring l1 and l2 norms? If not, can we come up with better names so that they are not confused with l1 and l2 norms?

RichDavis1 replied:
These are in reference to l1 and l2 norms.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+        $wordIndices = $this->approximation->wordIndices($predictWord);
+        $l2 = $this->layerMatrix($wordIndices);
+        $fa = $this->propagateHidden($l2, $l1);
+        $gd = $this->approximation->gradientDescent($fa, $predictWord['word'], $this->alpha);
```
What is returned by the `gradientDescent()` method? A vector of partial derivatives? If so, can we name the variable assigned to this data something more descriptive such as `$gradient`?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+     *
+     * @param string[] $sentences
+     */
+    private function preprocess(array $sentences) : void
```
Can any of this preprocessing be done outside of fitting in a preprocessing step?

RichDavis1 replied:
The question is do we handle pre-processing in the constructor or in fit?

The "preprocess" method handles all vocabulary preprocessing so we can move it wherever we need to. The reason it's currently in the fit method is because the fit contract requires the Dataset parameter. So we don't have the information required to preprocess until the fit method is called.

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+    private function prepareWeights() : void
+    {
+        for ($i = 0; $i < $this->vocabCount; ++$i) {
+            $this->syn1[] = Vector::zeros($this->dimensions);
```
Could the zero vector be instantiated outside of this loop for performance reasons?

andrewdalpino commented on src/Embedders/Word2Vec.php:
```diff
@@ -0,0 +1,790 @@
+        foreach ($sentence as $word) {
+            $vocabItem = $this->vocab[$word] ?? false;
+
+            if (!empty($vocabItem) && $vocabItem['sample_int'] > ($rand * self::RAND_MULTIPLIER)) {
```
Can the value of `$rand * self::RAND_MULTIPLIER` be assigned to a variable and then used here instead so we don't have to unnecessarily keep multiplying a static variable with a constant?

andrewdalpino commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+ * is greater than or equal to the key of C. In a min heap, the key of P is less than or equal to the key of C.
+ *
+ * References:
+ * [1] Black (ed.), Paul E. (2004-12-14)
```
We generally use the format `Authors. (Year). Title of paper.` in our code citations

andrewdalpino commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+class Heap
+{
+    /**
+     * An array of words from a corpus arranged in a manner that maintains the heap invariant.
```
'that maintains the (min or max?) heap invariant.'

andrewdalpino replied:
Why are both the class name and this variable named 'heap'? Is there a more specific name we could give this variable?

RichDavis1 replied:
Are you referring to the heap property?

The heap is well, the heap. It's the specific object the class creates. 

There is no min or max heap invariant, here is the definition: https://docs.python.org/3/library/heapq.html

andrewdalpino replied:
I am referring to the name you chose to give to this variable ... Could we name it something other than heap since we already called the aggregate root a "heap".

RichDavis1 replied:
I can't see which variable you're referring to in the snippet for this comment. It just shows the Heap class.

andrewdalpino commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+ * @package     RubixML
+ * @author      Rich Davis
+ */
+class Heap
```
What kind of heap is this? A min heap? A max heap? Some other heap?

RichDavis1 replied:
This is a heap queue, more commonly known as a priority queue, there is no min or max heap. 

If we're referring to how the pop method operates, it will return the smallest item, not the largest, this is what most refer to as a min heap which we can reference.

andrewdalpino replied:
So how does the class name HeapQueue sound?

RichDavis1 replied:
That's fine.

andrewdalpino commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+     */
+    public function __construct(array $x)
+    {
+        if (!is_array($x)) {
```
An array is typehinted in the constructor? Is this check necessary?

andrewdalpino commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+            throw new InvalidArgumentException('Heap requires a valid array.');
+        }
+
+        $this->heap = $this->heapify($x);
```
It looks like this method is used only in the constructor and is not overriden, why not just in-line this logic in the constructor?

RichDavis1 replied:
Wanted to abstract as that seems like too much logic handled in the constructor.

andrewdalpino commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+     *
+     * @return array[]
+     */
+    public function heap() : array
```
What value does this method provide to the end user?

RichDavis1 replied:
In the doc block it reads "returns the heap to the user". This will return the heap property so should probably be renamed to "getHeap".

andrewdalpino replied:
Why would the user need to access this heap? Is this method used anywhere else in your code? If not, and it only provides value in the edge case, we should consider removing it as we will have to maintain it as part of our public API.

RichDavis1 replied:
No but we do need to get the heap object in HierarchicalSoftmax hence the public visibility.

andrewdalpino reviewed (commented):
Hey @RichDavis1 looking better so far ... I'm submitting the comments that I have so far because we now have a 0.2.0 branch in the main repo and I'd like to transfer this PR and target that branch. Can you take a look at what I commented on and then submit a PR to the main repo with the changes?

simplechris commented on src/Embedders/SoftmaxApproximations/HierarchicalSoftmax.php:
```diff
@@ -0,0 +1,128 @@
+     */
+    public function structureSampling(Word2Vec $word2vec) : void
+    {
+        if (!$word2vec instanceof Word2Vec) {
```
This is already enforced by the type hint above

simplechris commented on src/Embedders/SoftmaxApproximations/HierarchicalSoftmax.php:
```diff
@@ -0,0 +1,128 @@
+        }
+
+        $this->vocab = $word2vec->vocab();
+        $vocabCount = $word2vec->vocabCount();
```
should this be using the instance variable at `$this->vocabCount` or can that be removed?

RichDavis1 replied:
We do this in order to reduce how many times we count the vocab. As it pertains to w2v and php, shaving off any redundant processes adds up.

simplechris commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+ * @package     RubixML
+ * @author      Rich Davis
+ */
+class Heap
```
I think this datastructure implementation could be refactored to implement `\Iterator` , `\Countable` interfaces so operations can be performed directly on this object and not just exposing the internal array structure.

Come to think of it, could you just extend `SplMinHeap`? and use the `insert`/`extract` semantics in places of the `heappush`/`heappop`?

See: https://www.php.net/manual/en/class.splheap.php

RichDavis1 replied:
That had been tried but we need to maintain the reference to the vocab item's index, count, code, and point items. This is why the heap class had to be implemented. There was no other way to match everything back up after the queue was prioritized.

simplechris commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+     *
+     * @return mixed[]|null
+     */
+    public function heappop() : ?array
```
This feels more synonymous with an `extract` operation on a heap structure

RichDavis1 replied:
Since this a priority queue, "pop" will return the smallest item and while maintaining the heap invariant. "Pop" and "extract" are somewhat synonymous but we would have to explicitly state "extract-min" in this situation. I referenced python's priority queue naming convention which reads "heappop".

```
heapq.heappop(heap)
Pop and return the smallest item from the heap, maintaining the heap invariant. If the heap is empty, IndexError is raised. To access the smallest item without popping it, use heap[0].
```

https://docs.python.org/3/library/heapq.html

simplechris replied:
ah gotcha, handn't grokked initially that this was a priority queue, and not a generic (min|max)heap structure 👍

simplechris commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+     * @param mixed[] $item
+     * @return mixed[]
+     */
+    public function heappush(array $item) : array
```
<s>...and I think that this would typically be an `insert` operation</s> 

**EDIT:** retracted

RichDavis1 replied:
See above, for reference: https://docs.python.org/3/library/heapq.html

```
heapq.heappush(heap, item)
Push the value item onto the heap, maintaining the heap invariant.
```

andrewdalpino replied:
The choice of verbs does not bother me, but why do we need to repeat the name heap in these methods when the object is called a heap?

simplechris replied:
I now agree that `pop`/`push` are decent semantics (and retract my earlier `insert`/`extract` suggestion). 

I totally understand that the current naming is in keeping with the reference implementation, but agree that `Heap::heappush(array $item) : array` could be renamed `Heap::push(array $item) : array` to be cleaner and not result in any loss of meaning 👍

RichDavis1 replied:
Updated

simplechris commented on src/Graph/Trees/Heap.php:
```diff
@@ -0,0 +1,179 @@
+     * @param array[] $x
+     * @return mixed[]
+     */
+    private function heapify(array $x) : array
```
This logic could live in the constructor 👍

RichDavis1 replied:
Mentioned in comment: https://github.com/RubixML/Extras/pull/1#discussion_r468221782

Marking resolved to keep conversation in a centralized place.

simplechris reviewed (commented):
Great stuff, @RichDavis1! awesome enhancement.

I see @andrewdalpino has provided some great pointers and thought i'd add a few general comments as I read through it for the first time and familiarized myself with this implementation. 👍

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Edit Word2Vec.php, NegativeSampling.php, HierarchicalSoftmax.php, SoftmaxApproximator.php, Heap.php

Addressed outstanding comments, cleaned up redundant processes, and updated several naming conventions.

andrewdalpino commented:
Hey @RichDavis1 we were looking good until I realized you unresolved alot of comments without actually making the changes in code. Also thanks @simplechris for checking out the Heap code!

andrewdalpino commented on tests/Embedders/Word2VecTest.php:
```diff
@@ -0,0 +1,141 @@
+     *
+     * @var int
+     */
+    protected const DATASET_SIZE = 2;
```
Do we need this constant?

RichDavis1 replied:
Removed

andrewdalpino reviewed (changes requested):
Hey @RichDavis1 we were looking good until I realized you unresolved some of comments without responding or making changes. Also thanks @simplechris for checking out the Heap code!

RichDavis1 committed:
Edit SoftmaxApproximations namespace, edit Word2Vec.php, edit Heap.php

Renamed SoftmaxApproximations namespace & interface to SoftmaxApproximator, renamed Heap to Heapqueue, renamed several HeapQueue methods, removed unnecessary data type validations.

RichDavis1 committed:
Merge branch '0.3.0-beta'

RichDavis1 committed:
Edit HierarchicalSoftmax.php, Edit Word2vec.php, edit HeapQueue.php, edit Word2VecTest.php

Linting changes.

andrewdalpino commented:
Hey @RichDavis1 can you please sign the CLA so we can get this merged? Thank you

RichDavis1 commented:
recheck

RichDavis1 commented:
I have read the CLA Document and I hereby sign the CLA

andrewdalpino commented:
Crap it deleted the PR when I merged 0.3.0-beta into master ... can you retarget master? Im sorry bro @RichDavis1 one thing after another but I think we're through the woods after this

andrewdalpino commented:
@RichDavis1 I had to pull this PR out of master because I forgot to squash your commits before merging. I don't know how to get it to "resubmit" so I can squash and merge this time. Can you do me a huge favor and submit the PR again from your end?

Don't worry about updating it to ML 1.0 (unless you want to) I can take care of that. Thanks man, sorry lol
